OBFrom 1d1b16066fe3f75a178c9aad324b7200f0ceafa2 Mon Sep 17 00:00:00 2001
From: Maximilian Bosch <maximilian.bosch@mayflower.de>
Date: Sat, 12 Aug 2023 09:42:55 +0000
Subject: [PATCH 2/2] Implement superuser-to-oidc mapping similar to the
 approach from LDAP auth

With LDAP auth it was possible to define a `superuser group`, i.e.
each member of that group became a superuser in mailman. This patch
replicates this behavior with OIDC authentication:

* If no `groups`-output is part of the response (i.e. `extra_data`),
  nothing happens - as before.
* Otherwise it's checked if the user is member of a hardcoded `admin`
  group. And depending on whether or not, the superuser flag in `auth_user`
  is true or false.
---
 allauth/account/utils.py                      | 7 +++++++
 allauth/socialaccount/internal/flows/login.py | 4 +++-
 2 files changed, 10 insertions(+), 1 deletion(-)

diff --git a/allauth/account/utils.py b/allauth/account/utils.py
index 0d0f6cb8..7afddb05 100644
--- a/allauth/account/utils.py
+++ b/allauth/account/utils.py
@@ -297,6 +297,13 @@ def filter_users_by_username(*username):
     return ret
 
 
+def _update_superuser_status(sociallogin):
+    user = _find_existing_user_by_email(user_email(sociallogin.user))
+    is_superuser = 'admin' in sociallogin.account.extra_data.get('groups', [])
+    user.is_superuser = is_superuser
+    user.save()
+
+
 def _find_existing_user_by_email(email):
     """Searches for an existing user by the email and returns the user model."""
     email_field = account_settings.USER_MODEL_EMAIL_FIELD
diff --git a/allauth/socialaccount/internal/flows/login.py b/allauth/socialaccount/internal/flows/login.py
index df50ae6d..f85c4e94 100644
--- a/allauth/socialaccount/internal/flows/login.py
+++ b/allauth/socialaccount/internal/flows/login.py
@@ -3,7 +3,7 @@ from django.shortcuts import render
 
 from allauth.account import app_settings as account_settings
 from allauth.account.adapter import get_adapter as get_account_adapter
-from allauth.account.utils import perform_login
+from allauth.account.utils import perform_login, _update_superuser_status
 from allauth.core.exceptions import (
     ImmediateHttpResponse,
     SignupClosedException,
@@ -78,9 +78,11 @@ def _authenticate(request, sociallogin):
     if sociallogin.is_existing:
         # Login existing user
         ret = _login(request, sociallogin)
+        _update_superuser_status(sociallogin)
     else:
         # New social user
         ret = process_signup(request, sociallogin)
+        _update_superuser_status(sociallogin)
     return ret
 
 
